consensus/src/consensus/inner/mod.rs

```diff
@@ -0,0 +1,192 @@
+
+// The snarkOS library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The snarkOS library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the snarkOS library. If not, see <https://www.gnu.org/licenses/>.
+
+use std::sync::Arc;
+
+use crate::{
+    error::ConsensusError,
+    Consensus,
+    ConsensusParameters,
+    CreatePartialTransactionRequest,
+    DeserializedLedger,
+    DynLedger,
+    MemoryPool,
+    Tx,
+};
+use anyhow::*;
+use snarkos_storage::{
+    BlockStatus,
+    Digest,
+    ForkDescription,
+    SerialBlock,
+    SerialTransaction,
+    Storage,
+    VMRecord,
+    VMTransaction,
+};
+use snarkvm_dpc::{
+    testnet1::{
+        instantiated::{Components, InstantiatedDPC},
+        payload::Payload as RecordPayload,
+        transaction::amount::AleoAmount,
+        Record as DPCRecord,
+        TransactionKernel,
+    },
+    DPCScheme,
+};
+use snarkvm_posw::txids_to_roots;
+use tokio::sync::mpsc;
+
+use snarkos_metrics::misc::*;
+
+use rand::thread_rng;
+
+use super::message::{ConsensusMessage, CreateTransactionRequest, TransactionResponse};
+
+mod agent;
+mod commit;
+mod transaction;
+
+pub struct ConsensusInner {
+    pub public: Arc<Consensus>,
+    pub ledger: DynLedger,
+    pub memory_pool: MemoryPool,
+    pub storage: Arc<dyn Storage>,
+}
+
+struct LedgerData {
+    ledger: DynLedger,
+    commitments: Vec<Digest>,
+    serial_numbers: Vec<Digest>,
+    memos: Vec<Digest>,
+    ledger_digests: Vec<Digest>,
+}
+
+impl ConsensusInner {
+    /// scans uncommitted blocks for forks
+    async fn scan_forks(&mut self) -> Result<()> {
+        let blocks = self
+            .storage
+            .get_block_hashes(None, snarkos_storage::BlockFilter::NonCanonOnly)
+            .await?;
+        info!("scanning {} blocks for forks", blocks.len());
+        for hash in blocks {
+            let header = self.storage.get_block_header(&hash).await?;
+            let parent_state = self.storage.get_block_state(&header.previous_block_hash).await?;
+            match parent_state {
+                BlockStatus::Unknown => continue, // orphan
+                BlockStatus::Committed(_) => (),  // uncomitted child of canon, could be a fork or a hanging block
+                BlockStatus::Uncommitted => {
+                    // mid-fork or orphan chain, wait for head of fork
+                    continue;
+                }
+            };
+            let block = self.storage.get_block(&hash).await?;
+
+            self.try_commit_block(&hash, &block).await?;
```

Comment:
are these blocks sorted by their parent-child relationship, i.e. we are likely to process these blocks in the correct order so they can be committed?